### ISSUE TITLE
Carrier not fully blocked in CW_L, CW_U, USB and LSB

### ISSUE THREAD

**Issue by dd0vs** [kb1vc/SoDaRadio#10]
Hallo Matt,
during observing SoDaRadio with SDRPlay, it was found that during CW and SSB with out Modulation or keying a carrier is send out.
Can this be blocked somehow?
vy73
Harald
DD0VS

**Comment by kb1vc** [comment]
How strong is the carrier?  

I believe that the best we can do for suppression is about 40dB or so.  

Are you using a B2xx series?  N2xx?  The N2xx and others can use the uhd_cal_tx_iq_balance
program to optimize the suppression.  The B200 cannot, as it lacks a "CAL" path from its
own TX to RX (and there may be other issues).  On the other hand, perhaps this should just
be built in to SoDaRadio.  

The issue here, as with all SDRs is that the "quadrature" oscillators in the signal chain are never quite perfectly in quadrature.  But small angle errors can result in perceptible carrier or sideband energy.  

I ought to look at this.  I'm surprised you're seeing carrier out with *no* modulation, since I specifically tested both the B210 and N200/UBX for that.   I'll take a look at this over the weekend. 

In the mean time, how strong is the carrier?  (relative to output with modulation.)  

73, 
matt

**Comment by dd0vs** [comment]
Hello,
I will try to give more specific info. But in CW I can listen 2 tones.

Vy73
Harald 
DD0VS 

.-.-. --... ...--     -.. -.. ----- ...- ...

**Comment by dd0vs** [comment]
I forgot, I have a B200mini.


.-.-. --... ...--     -.. -.. ----- ...- ...

**Comment by dd0vs** [comment]
Hello,
sorry for 3 mails in one chain.
See Pictures attached in CW this carriers is About -80dbm and the tone -55dBm.
In SSB -80dBm and the strongest tone is -60dBm.
(taken with RX: SDRPlay1, SDR-Radio V3 and HB9CV – TX: B200mini + 50Ohm Dipol of F9FT – Distance 1m – i can do it without antenna an with a atentuator on the Weekend hopefully). 

Gesendet von Mail für Windows 10

Von: kb1vc
Gesendet: Mittwoch, 6. Mai 2020 06:00
An: kb1vc/SoDaRadio
Cc: dd0vs; Author
Betreff: Re: [kb1vc/SoDaRadio] Carrier not fully blocked in CW_L, CW_U, USBand LSB (#10)

How strong is the carrier?
I believe that the best we can do for suppression is about 40dB or so.
Are you using a B2xx series? N2xx? The N2xx and others can use the uhd_cal_tx_iq_balance
program to optimize the suppression. The B200 cannot, as it lacks a "CAL" path from its
own TX to RX (and there may be other issues). On the other hand, perhaps this should just
be built in to SoDaRadio.
The issue here, as with all SDRs is that the "quadrature" oscillators in the signal chain are never quite perfectly in quadrature. But small angle errors can result in perceptible carrier or sideband energy.
I ought to look at this. I'm surprised you're seeing carrier out with no modulation, since I specifically tested both the B210 and N200/UBX for that. I'll take a look at this over the weekend.
In the mean time, how strong is the carrier? (relative to output with modulation.)
73,
matt

**Comment by dkozel** [comment]
The B2xx series of USRPs have automatic IQ and DC correction loops running in the AD936x RFIC frontend.

Usually the answer is to use the offset tuning feature of the USRP to move the local oscillator leakage away from the transmitted or received tone. You're still generating it, but it's out of band of your main signal.
https://files.ettus.com/manual/page_general.html#general_tuning

**Comment by kb1vc** [comment]
Yup....  And that's what I do for the UBX on HF.  

I think I've got something that may be "not as bad" as it was.  Version 7.1.0 moves the TX RF oscillator up the band (or down, depending) and typically out-of-band.  On 2m, a 144.2 signal gets an RF LO of 150MHz.  (The setting ends up the nearest multiple of 12.5, 10, or 50MHz...) 

Harald, Could you try building from this version and see what happens? I took a look using the
full-duplex monitor and the in-band carrier has disappeared.  SSB seems to have done
the same.   In this case, "disappeared" means -60dBc, so there may still be some signal left. 
A 20W PA is going to apply 23dB +/- of gain.  Though 60dB down from 43dBm is not going to blow the neighbor's ears off, but it will be detectable. I'm hoping that being 6MHz out-of-band will give the filters some room to work.  If there are other bands you'd like to look at, please let me know.  

I wanted to get this out to you quickly, but realize that I'm asking you to be a test pilot.  I'll
understand if you can't though.   If this does work out, I'll make it part of the main branch.

```
commit 7692d729ba90fe137b2eadca294499d0cb490ad5 (HEAD -> move_tx_carrier, origin/move_tx_carrier)
Author: Matt Reilly <kb1vc@kb1vc.org>
Date:   Wed May 6 21:18:22 2020 -0400

    A user points out that the carrier is not well suppressed in CW mode
    when there is no keydown, or in SSB when there is no modulation.
    This is especially true of the B210.  Measured suppression is > 40dBc
    but this is not sufficient.
    
    This commit applies a target frequency correction to the TX frequency
    to push the transmit frequency at least 1MHz.

```

Derek -- good to hear from you.  I hope all is well 

73
matt

**Comment by kb1vc** [comment]
Wow, and what an awful commit message-- 

I'll fix that up if and when this gets merged. 

The intent is to move the transmit LO at least 1MHz away from the desired transmit frequency. 
Though currently the separation is 5+ MHz from 144.

**Comment by dd0vs** [comment]
Matt, All,
i apologize for the awfull title of #10.

I will do better in future.

Sorry
Harald
DD0VS

Gesendet von Mail für Windows 10

Von: kb1vc
Gesendet: Donnerstag, 7. Mai 2020 03:50
An: kb1vc/SoDaRadio
Cc: dd0vs; Author
Betreff: Re: [kb1vc/SoDaRadio] Carrier not fully blocked in CW_L, CW_U, USBand LSB (#10)

Wow, and what an awful commit message--
I'll fix that up if and when this gets merged.
The intent is to move the transmit LO at least 1MHz away from the desired transmit frequency.
Though currently the separation is 5+ MHz from 144.

**Comment by dd0vs** [comment]
Hello Matt,

i am not able to see Version 7.1 on GitHub?

Planning to do some test this evening.

Vy73
Harald
DD0VS

Gesendet von Mail für Windows 10

Von: kb1vc
Gesendet: Donnerstag, 7. Mai 2020 03:45
An: kb1vc/SoDaRadio
Cc: dd0vs; Author
Betreff: Re: [kb1vc/SoDaRadio] Carrier not fully blocked in CW_L, CW_U, USBand LSB (#10)

Yup.... And that's what I do for the UBX on HF.
I think I've got something that may be "not as bad" as it was. Version 7.1.0 moves the TX RF oscillator up the band (or down, depending) and typically out-of-band. On 2m, a 144.2 signal gets an RF LO of 150MHz. (The setting ends up the nearest multiple of 12.5, 10, or 50MHz...)
Harald, Could you try building from this version and see what happens? I took a look using the
full-duplex monitor and the in-band carrier has disappeared. SSB seems to have done
the same. In this case, "disappeared" means -60dBc, so there may still be some signal left.
A 20W PA is going to apply 23dB +/- of gain. Though 60dB down from 43dBm is not going to blow the neighbor's ears off, but it will be detectable. I'm hoping that being 6MHz out-of-band will give the filters some room to work. If there are other bands you'd like to look at, please let me know.
I wanted to get this out to you quickly, but realize that I'm asking you to be a test pilot. I'll
understand if you can't though. If this does work out, I'll make it part of the main branch.
commit 7692d729ba90fe137b2eadca294499d0cb490ad5 (HEAD -> move_tx_carrier, origin/move_tx_carrier)
Author: Matt Reilly <kb1vc@kb1vc.org>
Date:   Wed May 6 21:18:22 2020 -0400

    A user points out that the carrier is not well suppressed in CW mode
    when there is no keydown, or in SSB when there is no modulation.
    This is especially true of the B210.  Measured suppression is > 40dBc
    but this is not sufficient.
    
    This commit applies a target frequency correction to the TX frequency
    to push the transmit frequency at least 1MHz.

Derek -- good to hear from you. I hope all is well
73
matt

**Comment by dd0vs** [comment]
Hello,
i confirm, that the branch "move_tx_carrier" solves the raised issue, sorry it took a while till i found it.
Its understood that this is no a carrier, but local oscilator bleed through.
move_tx_carrier move this away, but understanding is, that that must be filtered, even when power attentuation is about 40dB.
I whish to have a bit more control about the LO frequency to ensure it is outside a filter passband.

Many thanks for understanding and helping!
vy73
Harald
DD0VS

**Comment by kb1vc** [comment]
Great.  I'll close this issue then.   (Finally getting back to things that have been sitting for a long while. )

And my apologies-- I did not mean that *your* message was horrible, I was commenting on my own message in the commit log.  
```
    This commit applies a target frequency correction to the TX frequency
    to push the transmit frequency at least 1MHz.
```

I generally try to be more specific.  Apparently I pushed the commit button before finishing the thought that started with "to push the transmit frequency at least 1MHz."  The goal was to push the analog LO well out of the passband from any reasonable transmit filters. 

And again, thank you for being a SoDaRadio user. 

73
